### PR TITLE
os: use transparent comparator in ObjectStore::getattrs()

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by rzarzynski — APPROVED** [review]

**Comment by ronen-fr** [comment]
Is there a meaningful name we can use for this recurring type (map<string,bufferptr,less<>>)?

File: src/crimson/os/cyanstore/cyan_store.cc [hunk 1]
```diff
@@ -405,7 +405,7 @@ seastar::future<> CyanStore::do_transaction(CollectionRef ch,
       {
         coll_t cid = i.get_cid(op->cid);
         ghobject_t oid = i.get_oid(op->oid);
-        std::map<std::string, ceph::bufferlist> aset;
+        std::map<std::string, ceph::bufferptr> aset;
```
Comment (ronen-fr) [inline src/crimson/os/cyanstore/cyan_store.cc:408]: the comparator was left out intentionally? (here and below)

Comment (tchaikov) [reply src/crimson/os/cyanstore/cyan_store.cc:408]: @ronen-fr i am still working on the crimson side.

**Review by ronen-fr — APPROVED** [review]

**Comment by tchaikov** [comment]
@rzarzynski @ronen-fr i dropped the change to replace bufferlist with bufferptr in `FuturizedStore::get_attrs()`. despite that it saves us the overhead of boxing a memory chunk to a bufferlist in the i/o path, and it's aligned with the type of memory chunk returned from seastore and alienstore, it'd introduce much more code churn than this change alone.

i will create another PR to do this change.

**Comment by tchaikov** [comment]
> Is there a meaningful name we can use for this recurring type (map<string,bufferptr,less<>>)?

probably, it depends on the context.

File: src/crimson/os/alienstore/alien_store.cc [hunk 2]
```diff
@@ -338,13 +338,7 @@ AlienStore::get_attrs(CollectionRef ch,
...
     return tp->submit(ch->get_cid().hash_to_shard(tp->size()), [=, &aset] {
       auto c = static_cast<AlienCollection*>(ch.get());
-      std::map<std::string, ceph::bufferptr> blueaset;
-      const auto r = store->getattrs(c->collection, oid, blueaset);
-      for (auto& [bluekey, blueval] : blueaset) {
-        ceph::bufferlist bl;
-        bl.push_back(std::move(blueval));
-        aset.emplace(std::move(bluekey), std::move(bl));
-      }
+      const auto r = store->getattrs(c->collection, oid, aset);
```
Comment (rzarzynski) [inline src/crimson/os/alienstore/alien_store.cc:341]: Hmm, that's stange. IIUC `attrs_t` is still about `bufferlist` while the sole change for `BlueStore` is just about the comparator:

```diff
--- a/src/os/bluestore/BlueStore.cc
+++ b/src/os/bluestore/BlueStore.cc
@@ -10431,7 +10431,7 @@ int BlueStore::getattr(
 int BlueStore::getattrs(
   CollectionHandle &c_,
   const ghobject_t& oid,
-  map<string,bufferptr>& aset)
+  map<string,bufferptr,less<>>& aset)
 {
   Collection *c = static_cast<Collection *>(c_.get());
   dout(15) << __func__ << " " << c->cid << " " << oid << dendl;
diff --git a/src/os/bluestore/BlueStore.h b/src/os/bluestore/BlueStore.h
index a1249607a4d2a..c08cc6d61aac8 100644
--- a/src/os/bluestore/BlueStore.h
+++ b/src/os/bluestore/BlueStore.h
@@ -2836,7 +2836,7 @@ class BlueStore : public ObjectStore,
 	      ceph::buffer::ptr& value) override;
 
   int getattrs(CollectionHandle &c, const ghobject_t& oid,
-	       std::map<std::string,ceph::buffer::ptr>& aset) override;
+	       std::map<std::string,ceph::buffer::ptr, std::less<>>& aset) override;
 
   int list_collections(std::vector<coll_t>& ls) override;
 ```

If so, how would it pass the `make check` bot?

Comment (tchaikov) [reply src/crimson/os/alienstore/alien_store.cc:341]: @rzarzynski there are two variants of `getattrs()`. one for bufferptr and the other for bufferlist. before this change, the variant which returns `std::map<std::string, ceph::bufferptr>` is used. after this change the variant returning `std::map<std::string, ceph::bufferlist>` is used. so we don't need to box the bufferptr as bufferlist here manually.

Comment (rzarzynski) [reply src/crimson/os/alienstore/alien_store.cc:341]: Ah, I see. Thanks.

**Review by rzarzynski — APPROVED** [review]

**Review by liu-chunmei — APPROVED** [review]

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-06-29_13:42:32-rados-wip-kefu-testing2-2021-06-29-1946-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-06-30_04:57:30-rados-wip-kefu-testing2-2021-06-30-1019-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-06-30_08:47:33-rados-wip-kefu-testing2-2021-06-30-1405-distro-basic-smithi/

**Pull request merged**